sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionServiceSchemaTransformProvider.java

```diff
@@ -0,0 +1,174 @@
+        PCollectionTuple inputTuple = (PCollectionTuple) input;
+        for (TupleTag<?> tag : inputTuple.getAll().keySet()) {
+          inputRowTuple = inputRowTuple.and(tag.getId(), (PCollection<Row>) inputTuple.get(tag));
+        }
+      } else {
+        throw new RuntimeException(String.format("Unsupported input type: %s", input));
+      }
+      PCollectionRowTuple output = inputRowTuple.apply(this.rowTuplePTransform);
+
+      if (output.getAll().size() > 1) {
+        PCollectionTuple pcTuple = PCollectionTuple.empty(input.getPipeline());
+        for (String key : output.getAll().keySet()) {
+          pcTuple = pcTuple.and(key, output.get(key));
+        }
+        return pcTuple;
+      } else if (output.getAll().size() == 1) {
+        return output.getAll().values().iterator().next();
+      } else {
+        return PDone.in(input.getPipeline());
+      }
+    }
+  }
+
+  @Override
+  public PTransform getTransform(FunctionSpec spec) {
+    SchemaTransformPayload payload;
+    try {
+      payload = SchemaTransformPayload.parseFrom(spec.getPayload());
+      String identifier = payload.getIdentifier();
+      if (!schemaTransformProviders.containsKey(identifier)) {
+        throw new RuntimeException(
+            "Did not find a SchemaTransformProvider with the identifier " + identifier);
+      }
+
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(
+          "Invalid payload type for URN " + getUrn(ExpansionMethods.Enum.SCHEMATRANSFORM), e);
+    }
+
+    String identifier = payload.getIdentifier();
+    org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider provider =
+        schemaTransformProviders.get(identifier);
+    if (provider == null) {
+      throw new IllegalArgumentException(
+          "Could not find a SchemaTransform with identifier " + identifier);
+    }
+
+    Schema configSchemaFromRequest =
+        SchemaTranslation.schemaFromProto((payload.getConfigurationSchema()));
+    Schema configSchemaFromProvider = provider.configurationSchema();
+
+    if (!configSchemaFromRequest.assignableTo(configSchemaFromProvider)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Config schema provided with the expansion request %s is not compatible with the "
+                  + "config of the Schema transform %s.",
+              configSchemaFromRequest, configSchemaFromProvider));
+    }
+
+    Row configRow;
+    try {
+      configRow =
+          RowCoder.of(provider.configurationSchema())
+              .decode(payload.getConfigurationRow().newInput());
+    } catch (IOException e) {
+      throw new RuntimeException("Error decoding payload", e);
+    }
+
+    return new RowTransform(provider.from(configRow).buildTransform());
+  }
+
+  Iterable<org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider> getAllProviders() {
+    return schemaTransformProviders.values();
```

Comment:
so if a provider is not (duh) provided, even if the SchemaTransform implementation exists, it would not show up in the Discover call. Is that fine?

Reply:
I would hope that we can eventually replace writing a Provider with a decorator on the (suitably configured) PTransform class itself, but that's future work.